src/FHIR/SMART/SmartLaunchController.php

```diff
@@ -0,0 +1,155 @@
+            // appear to be required in the spec.
+
+            $gbl = \RestConfig::GetInstance();
+            // TODO: adunsulag surely we can centralize where this fhir API url is set?
+            // TODO: adunsulag what is wrong with these URL's?  I'm having to hard code the issuer as I can't
+            // seem to get these URLs right.  for some reason the $SITE is set to interface, we don't get 'apis' in there
+            // ROOT_URL appears to be empty.. just strange
+            // $issuer = $GLOBALS['site_addr_oath'] . $gbl::$SITE . $gbl::$ROOT_URL . "/fhir";
+            $issuer = $GLOBALS['site_addr_oath'] . "/apis/default/fhir";
+            $launchParams = "launch.html?launch=" . $launchCode . "&iss=" . $issuer;
+
+            expand_collapse_widget(
+                $widgetTitle,
+                $widgetLabel,
+                $widgetButtonLabel,
+                $widgetButtonLink,
+                $widgetButtonClass,
+                $linkMethod,
+                $bodyClass,
+                $widgetAuth,
+                $fixedWidth,
+                $forceExpandAlways
+            );
+            ?>
+            <div>
+                <ul>
+                        <?php if (empty($smartClients)) : ?>
+                            <li><p><?php echo xlt("No registered SMART apps in the system"); ?></p></li>
+                        <?php endif; ?>
+                        <?php foreach ($smartClients as $client) : ?>
+                            <li class="summary_item">
+                                <button class='btn btn-primary btn-sm smart-launch-btn' data-smart-name="<?php echo $client->getName(); ?>"
+                                        data-smart-redirect-url="<?php echo $client->getRedirectUri() . $launchParams; ?>">
+                                    Launch
+                                </button>
+                                <?php echo $client->getName(); ?>
+                            </li>
+                        <?php endforeach; ?>
+                </ul>
+            </div>
+        </section>
+        <?php
+        // it's too bad we don't have a centralized page renderer we could tie this into and render javascript at the
+        // end of our footer pages on everything...
+        ?>
+        <script>
+            (function(window) {
+                let smartLaunchers = document.querySelectorAll('.smart-launch-btn');
+                for (let launch of smartLaunchers) {
+                    let url =
+                        launch.addEventListener('click', function(evt) {
+                            let node = evt.target;
+                            let url = node.dataset.smartRedirectUrl;
+                            if (!url) {
+                                return;
+                            }
+                            let title = node.dataset.smartTitle || "<?php echo xlt("Smart App"); ?>";
```

Comment:
Where does node.dataset.smartTitle come from? Just making sure xss not an issue

Reply:
This is actually a bug and should be node.dataset.smartName.  

If you were asking how that attribute is generated, it comes from the html attribute on the button **data-smart-name**  The browser for all HTML5 doctypes automatically grabs anything with a **data-** attribute and stuffs it into the dataset DOM property.  The attributes are camelCased.

Since I'm now escaping those values based on your suggestions with **attr()** there shouldn't be an XSS problem here.

Reply:
it's such a relief that Brady gets to pick on someone else besides me for escaping.